diff --git a/timsort.h b/timsort.h
index 383584f..faf2232 100644
--- a/timsort.h
+++ b/timsort.h
@@ -38,6 +38,8 @@
 #include <stdint.h>
 #elif defined(_WIN32)
 typedef unsigned __int64 uint64_t;
+#elif defined(NeXT)
+typedef unsigned long long uint64_t;
 #endif
 
 #ifndef SORT_NAME
diff --git a/xmlIO.c b/xmlIO.c
index 007144c..afb4aa6 100644
--- a/xmlIO.c
+++ b/xmlIO.c
@@ -10,6 +10,10 @@
 
 #define IN_LIBXML
 #include "libxml.h"
+#ifdef NeXT
+#include <sys/param.h>
+extern char *getwd(char *);
+#endif
 
 #include <string.h>
 #include <stddef.h>
@@ -3762,7 +3766,11 @@ xmlOutputBufferFlush(xmlOutputBufferPtr out) {
 char *
 xmlParserGetDirectory(const char *filename) {
     char *ret = NULL;
+#ifdef NeXT
+    char dir[MAXPATHLEN + 1];
+#else
     char dir[1024];
+#endif
     char *cur;
 
 #ifdef _WIN32_WCE  /* easy way by now ... wince does not have dirs! */
@@ -3792,7 +3800,11 @@ xmlParserGetDirectory(const char *filename) {
 	else *cur = 0;
 	ret = xmlMemStrdup(dir);
     } else {
+#ifdef NeXT
+        if (getwd(dir) != NULL) {
+#else
         if (getcwd(dir, 1024) != NULL) {
+#endif
 	    dir[1023] = 0;
 	    ret = xmlMemStrdup(dir);
 	}
