--- a/src/luaconf.h
+++ b/src/luaconf.h
@@ -10,6 +10,12 @@
 
 #include <limits.h>
 #include <stddef.h>
+
+#ifdef NeXT
+#define LUA_USE_MKSTEMP
+#define LUA_USE_ISATTY
+#define LUA_USE_POPEN
+#endif
 
 
 /*
@@ -667,7 +673,11 @@
 */
 #if defined(LUA_USE_POPEN)
 
+#ifdef NeXT
+#define lua_popen(L,c,m) ((void)L, fflush(stdout), fflush(stderr), popen(c,m))
+#else
 #define lua_popen(L,c,m)	((void)L, fflush(NULL), popen(c,m))
+#endif
 #define lua_pclose(L,file)	((void)L, (pclose(file) != -1))
 
 #elif defined(LUA_WIN)
--- a/src/loslib.c
+++ b/src/loslib.c
@@ -18,6 +18,10 @@
 
 #include "lauxlib.h"
 #include "lualib.h"
+
+#if defined(NeXT) && !defined(CLOCKS_PER_SEC)
+#define CLOCKS_PER_SEC CLK_TCK
+#endif
 
 
 static int os_pushresult (lua_State *L, int i, const char *filename) {
--- a/src/lvm.c
+++ b/src/lvm.c
@@ -206,7 +206,11 @@
   const char *r = getstr(rs);
   size_t lr = rs->tsv.len;
   for (;;) {
+#ifdef NeXT
+    int temp = strcmp(l, r);
+#else
     int temp = strcoll(l, r);
+#endif
     if (temp != 0) return temp;
     else {  /* strings are equal up to a `\0' */
       size_t len = strlen(l);  /* index of first `\0' in both strings */
